top-level: Create `pkgs{Build,Host,Target}{Build,Host,Target}`

Ericson2314 opened the pull request:
##### Motivation for this change

###### top-level: Create `pkgs{Build,Host,Target}{Build,Host,Target}`

This is needed to avoid confusing and repeated boilerplate for
`fooForTarget`.  The vast majority of use-cases can still use
`buildPackages or `targetPackages`, which are now defined in terms of
these.

###### ghc, go, guile: Use new `pkgs*`

`pkgsBuildTarget` allows us to avoid repeated and confusing conditions.
The others merely provide clarity for one the foreign package set's
target platform matters.

##### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Ericson2314 committed:
top-level: Create `pkgs{Build,Host,Target}{Build,Host,Target}`

This is needed to avoid confusing and repeated boilerplate for
`fooForTarget`.  The vast majority of use-cases can still use
`buildPackages or `targetPackages`, which are now defined in terms of
these.

Ericson2314 committed:
ghc, go, guile: Use new `pkgs*`

`pkgsBuildTarget` allows us to avoid repeated and confusing conditions.
The others merely provide clarity for one the foreign package set's
target platform matters.

matthewbauer commented on pkgs/top-level/stage.nix:
```diff
@@ -70,11 +75,32 @@ let
+
+    # Older names for package sets. Use these when only the host platform of the
+    # package set matter (i.e. use `buildPackages` where any of `pkgsBuild*`
+    # would do, and `targetPackages` when any of `pkgsTarget*` would do.)
```
There’s only one pkgsTarget* right.

Ericson2314 replied:
Yes. I should make that clear I'm saying would there be more than one it wouldn't matter.

matthewbauer commented on pkgs/top-level/stage.nix:
```diff
@@ -21,18 +21,23 @@
+  #     pkgsTargetTarget ...;
+  #   }
+  #
+  # These are references to adjacent bootstrapping stages. The more familiar
```
Maybe put this in Nixpkgs documentation.

matthewbauer commented on pkgs/development/compilers/ghc/8.2.2.nix:
```diff
@@ -70,11 +70,9 @@ let
-    else assert targetPlatform == hostPlatform; # build != host == target
-      [ stdenv.cc ] ++ stdenv.lib.optional useLLVM buildLlvmPackages.llvm;
+  toolsForTarget = [
+    pkgsBuildTarget.targetPackages.stdenv.cc
```
Why do you need targetPackages here?

Ericson2314 replied:
because `stdenv.cc` comes from the previous `stage. `targetPackages.stdenv` sort of cancels out so I get something like `pkgsBuildTarget.gcc` or `pkgsBuildTarget.llvm` in the end.

matthewbauer commented on pkgs/development/interpreters/guile/2.0.nix:
```diff
@@ -22,7 +22,7 @@
 
   depsBuildBuild = [ buildPackages.stdenv.cc ]
```
Should pkgsBuildBuild be used here?

Ericson2314 replied:
It would have to be `pkgsBuildBuild.targetPackages.stdenv.cc`, so ehh hardly worth it for now.

Mic92 commented on pkgs/top-level/stage.nix:
```diff
@@ -70,11 +75,32 @@ let
+    # platform. We only care about their host/target platforms, not their build
+    # platform, because the the former two alone affect the interface of the
+    # final package; the build platform is just an implementation detail that
+    # should not leak.
```
Please put also this into the documentation.

Ericson2314 committed:
manual: Document `pkgsFooBar` and more

There was a bunch of stuff in the cross section that haddn't had any
attention in a while. I might need to slim it down later, but this is
good for now.

Ericson2314 commented:
CC @shlevy here is that change the reduces the importance of the sliding window.

Ericson2314 commented on pkgs/stdenv/booter.nix:
```diff
@@ -95,13 +95,25 @@ stageFuns: let
+            if args.stdenv.hostPlatform == args.stdenv.targetPlatform
+            then thisStage
+            else assert args.stdenv.buildPlatform == args.stdenv.hostPlatform; pkgsBuildHost;
+          pkgsTargetTarget = nextStage;
```
Here is the proof while stages are "braided" not "chained" (more edges between them than `buildPackages` `targetPackages`), there are no more stages than before (same nodes).

Ericson2314 replied:
But if we ever had any stages that were canadian cross (build != host != target) then we would need more stages, as no {linear, sliding window}-generated stage would fit `pkgsBuildTarget` or `pkgsHostHost`. This relates to those two being conditionally hacked together above from the linear ones, while the rest are statically chosen.

Ericson2314 committed:
manual: Make sure building doesn't need recursive nix

$(shell ...) looks a little sketch like it will be run no matter what.
And there are problems building the manual on darwin so hopefully this
fixes them.

Ericson2314 committed:
manual: Auto reformat

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -12,11 +12,12 @@
-   that packages should be written with cross-compilation in mind, and nixpkgs
-   should evaluate in a similar way (by minimizing cross-compilation-specific
-   special cases) whether or not one is cross-compiling.
+   run-time environments! Significant, because the benefits apply even when one
```
```suggestion
   run-time environments!
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -12,11 +12,12 @@
-   should evaluate in a similar way (by minimizing cross-compilation-specific
-   special cases) whether or not one is cross-compiling.
+   run-time environments! Significant, because the benefits apply even when one
+   is developing and deploying on the same machine. Nixpkgs is increasingly
```
```suggestion
   Nixpkgs is increasingly
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -218,93 +219,119 @@
+    machine-code-producing tools, we're done. And indeed
+    <varname>buildInputs</varname> and <varname>nativeBuildInputs</varname>
+    have covered these simpler build-time and run-time (respectively) changes
+    for many years. But if the depedency does produce machine code, we might
```
```suggestion
    for many years. But if the dependency does produce machine code, we might
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -218,93 +219,119 @@
+    <varname>buildInputs</varname> and <varname>nativeBuildInputs</varname>
+    have covered these simpler build-time and run-time (respectively) changes
+    for many years. But if the depedency does produce machine code, we might
+    need to worry about it's target platform too. In principle, that target
```
```suggestion
    need to worry about its target platform too. In principle, that target
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -218,93 +219,119 @@
-    with native packages.
+    Finally, if the depending package is a compiler or other
+    machine-code-producing tool, it might need dependencies that run at "emit
+    time". This is for compilers that (regrettably) insist on being in built
```
```suggestion
    time". This is for compilers that (regrettably) insist on being built
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
+    linkend="ssec-cross-dependency-categorization"/> are specified as
+    lists of derivations given to <varname>mkDerivation</varname>, as
+    documented in <xref linkend="ssec-stdenv-dependencies"/>. In short, the
+    each list of dependencies for "host → target" of "foo → bar" is called
```
```suggestion
    the list of dependencies for "host → target" of "foo → bar" is called
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
+    <xref
+    linkend="ssec-cross-dependency-categorization"/> are specified as
+    lists of derivations given to <varname>mkDerivation</varname>, as
+    documented in <xref linkend="ssec-stdenv-dependencies"/>. In short, the
```
```suggestion
    documented in <xref linkend="ssec-stdenv-dependencies"/>. In short,
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
+    lists of derivations given to <varname>mkDerivation</varname>, as
+    documented in <xref linkend="ssec-stdenv-dependencies"/>. In short, the
+    each list of dependencies for "host → target" of "foo → bar" is called
+    <varname>depsFooBar</varname>, with the exceptions for backwards
```
```suggestion
    <varname>depsFooBar</varname>, with exceptions for backwards
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
+    <varname>buildPackages</varname>, <varname>pkgs</varname>, and
+    <varname>targetPackages</varname>. Those are now redefined as aliases to
+    <varname>pkgsBuildHost</varname>, <varname>pkgsHostTarget</varname>, and
+    <varname>pkgsTargetTarget</varname>. It is fine, indeed if anything
```
```suggestion
    <varname>pkgsTargetTarget</varname>. It is acceptable, even recommended
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
-  </note>
+
+   <para>
+    Now, for most of Nixpkgs's history, there was no
```
```suggestion
    Now, for most of Nixpkgs's history, there were no
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
+    their dependencies into the multiple lists described above. And indeed,
+    asking that users both sort their dependencies, <emphasis>and</emphasis>
+    take them from the right attribute set, is both too onerous and redundant,
+    so the recommend approach (for now) is to continue just categorizing by
```
```suggestion
    so the recommended approach (for now) is to continue just categorizing by
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
+   </para>
+
+   <para>
+    No make this work, we "splice" together the six
```
```suggestion
    To make this work, we "splice" together the six
```

matthewbauer commented on doc/submitting-changes.xml:
```diff
@@ -354,23 +354,22 @@ Additional information.
-     review changes from pull request number 12345:
-     <screen>nix-shell -p nix-review --run "nix-review pr 12345"</screen>
+    review changes from pull request number 12345:
+<screen>nix-shell -p nix-review --run "nix-review pr 12345"</screen>
```
?

Ericson2314 replied:
That is the autoformatter.

matthewbauer commented on doc/stdenv.xml:
```diff
@@ -222,9 +222,10 @@ genericBuild
-   could be enforced.
+   perfect sense outside of cross compilation. By default, the
+   run-time/build-time distinction is just a hint for mental clarity, but with
+   <varname>strictDeps</varname> set it is somewhat enforced even in the native
```
```suggestion
   <varname>strictDeps</varname> set it is enforced even in the native
```

Ericson2314 replied:
Well it's not *complete* enforcement. Maybe "mostly"?

Ericson2314 replied:
```suggestion
   <varname>strictDeps</varname> set it is mostly enforced even in the native
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
+    </para>
+    <para>
+     Second, it is because this is a frequent source of hard-to-follow
+     "infinite recursions" / cycles. When only packages sets that don't mention
```
```suggestion
     "infinite recursions" / cycles. When only package sets that don't mention
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
+    <para>
+     Second, it is because this is a frequent source of hard-to-follow
+     "infinite recursions" / cycles. When only packages sets that don't mention
+     target are used, the package set forms a directly acyclic graph. This
```
```suggestion
     target are used, the package set forms a directed acyclic graph. This
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
+     Second, it is because this is a frequent source of hard-to-follow
+     "infinite recursions" / cycles. When only packages sets that don't mention
+     target are used, the package set forms a directly acyclic graph. This
+     means that all cycles that exist are confirmed to one stage. This means
```
```suggestion
     means that all cycles that exist are confined to one stage. This means
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
+     "infinite recursions" / cycles. When only packages sets that don't mention
+     target are used, the package set forms a directly acyclic graph. This
+     means that all cycles that exist are confirmed to one stage. This means
+     they are a lot smaller, so easier to follow in the code or a backtrace. It
```
```suggestion
     they are a lot smaller, and easier to follow in the code or a backtrace. It
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
+     likely to be caught by CI and other users.
+    </para>
+    <para>
+     Thirdly, it is because everything target-mentioning only exists to
```
I disagree. A compiler will never be able to build everything, so the you will always need to keep track of target. Things like LLVM get close to compiling everything, but we might prefer to only build the part of LLVM we need.

Ericson2314 replied:
That's a good point. I think part of the answer is `targetPlatform` is less bad than `targetPackages`. I'll merge this now to unblock other stuff, but get back to fixing that wording later.

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
+     Thirdly, it is because everything target-mentioning only exists to
+     accommodate compilers with lousy build systems that insist on the compiler
+     itself and standard library being built together. Of course that is bad
+     because bigger derivation means longer rebuilds. It is also subpar because
```
```suggestion
     because bigger derivations means longer rebuilds. It is also problematic because
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
+    <filename>pkgs/top-level/stage.nix</filename> "ties the knot" (creates the
+    fixed point) of each stage. The package sets are defined per-stage however,
+    so they can be thought of as edges between stages (the nodes) in a graph.
+    Compositions like <literal>pkgsBuildTarget.TargetPackages</literal> can be
```
```suggestion
    Compositions like <literal>pkgsBuildTarget.targetPackages</literal> can be
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
+    <varname>pkgsHostTarget</varname> refers to the current one, and
+    <varname>pkgsTargetTarget</varname> refers to the next one. When there is
+    no previous or next stage, they instead refer to the current stage. Note
+    how all the invariants about the mapping between dependency and depending
```
```suggestion
    how all the invariants regarding the mapping between dependency and depending
```

matthewbauer commented on doc/cross-compilation.xml:
```diff
@@ -450,21 +477,202 @@ nix-build &lt;nixpkgs&gt; --arg crossSystem '{ config = "&lt;arch&gt;-&lt;os&gt;
+    <varname>pkgsBuildTarget</varname> and <varname>pkgsHostHost</varname> are
+    more complex in that the stage fitting the requirements isn't always a
+    fixed chain of "prevs" and "nexts" away (modulo the "saturating"
+    self-references at the ends). We just special case instead. All the primary
```
```suggestion
    self-references at the ends). We just special case each instead. All the primary
```

matthewbauer reviewed (changes requested):
grammar / edits / etc.

Ericson2314 committed:
manual: Fix typos

Thanks @matthewbauer!

Co-Authored-By: Ericson2314 <git@JohnEricson.me>
